### PR TITLE
Toa cleanup

### PR DESCRIPTION
**Pull request by aarchiba**
Substantial simplifications of the inner workings of `toa.py`; in particular abolishing `self.toas`.

- Now preserves the name column in tempo2 files
- make_fake_toas now uses ephemeris and other settings from the model
- get_TOAs can read and cache multiple .tim files
- pickling can be done manually with load_pickle and save_pickle
- TOAs can be checked against the files they were loaded from with check_hashes()
- TOAs can now be checked for equality with ==

### CODE REVIEW

**Comment by aarchiba** [comment]
I'm going to go over the docs but if someone would like to look at the code and tests, I'm not planning to add any major changes.

**Comment by aarchiba** [comment]
Ready to merge.

**Comment by aarchiba** [comment]
https://gitlab.nanograv.org/nano-time/timing_analysis/merge_requests/97 This makes it much easier to support transparent caching of NANOGrav data sets, desired for the v0 data release.

**Comment by paulray** [comment]
This looks like lots of good stuff. Is it the case now that `TOAs()` objects will always have a table ?

**Comment by aarchiba** [comment]
> This looks like lots of good stuff. Is it the case now that `TOAs()` objects will always have a table ?

Yes. They don't always have all the column one might wish - one has to call compute_TDB and some others fo have a fully "cooked" TOAs object, but they never contain lists any more.

**Comment by scottransom** [comment]
@aarchiba Does that affect performance in any way?  Positively or negatively?  It never bothered me that lists were an intermediate step because building things via lists is such a common Python idiom.  Curious if this makes things faster (or maybe even slower -- perhaps by expanding the Table)

**Comment by aarchiba** [comment]
> @aarchiba Does that affect performance in any way? Positively or negatively? It never bothered me that lists were an intermediate step because building things via lists is such a common Python idiom. Curious if this makes things faster (or maybe even slower -- perhaps by expanding the Table)

This has negligible performance impact - the Table is built from lists as before. The difference is that TOAs objects never contain lists, so functions don't need `if hasattr(self, "table")`.

**Comment by aarchiba** [comment]
In general this moves a lot of the construction/pickling code into helper functions, so that TOAs objects need to do less guesswork and type fiddling internally. Methods can assume that the object is reasonably complete and correct.

File: src/pint/toa.py [hunk 1]
```diff
@@ -142,14 +141,18 @@ def get_TOAs(
...
-    model : A valid PINT timing model or None
+    model : pint.models.timing_model.TimingModel or None
         If a valid timing model is passed, model commands (such as BIPM version,
         planet shapiro delay, and solar system ephemeris) that affect TOA loading
         are applied.
     usepickle : bool
         Whether to try to use pickle-based caching of loaded clock-corrected TOAs objects.
-    tdb_method : string
+    tdb_method : str
         Which method to use for the clock correction to TDB.
```
[the remark this answers is not in the document]
Comment (paulray) [reply src/pint/toa.py:151]: This is still used. The valid values are "default" and "ephemeris".

[the remark this answers is not in the document]
Comment (aarchiba) [reply src/pint/toa.py]: I linked to the observatory page that actually implements this.

**Pull request merged**